nwb_conversion_tools/utils/conversion_tools.py

```diff
@@ -1,67 +1,109 @@
+                    for data_interfaces in [global_data_interfaces, experiment_data_interfaces, session_data_interfaces]
+                    if data_interfaces is not None
+                ]
+            )
+            for data_interface_name in data_interfaces_names_chain:
+                data_interface_classes.update({data_interface_name: getattr(nwb_conversion_tools, data_interface_name)})
 
+            CustomNWBConverter = type(
+                "CustomNWBConverter", (NWBConverter,), dict(data_interface_classes=data_interface_classes)
+            )
 
-def make_nwbfile_from_metadata(metadata: dict):
-    """Make NWBFile from available metadata."""
-    metadata = dict_deep_update(get_default_nwbfile_metadata(), metadata)
-    nwbfile_kwargs = metadata["NWBFile"]
-    if "Subject" in metadata:
-        # convert ISO 8601 string to datetime
-        if "date_of_birth" in metadata["Subject"] and isinstance(metadata["Subject"]["date_of_birth"], str):
-            metadata["Subject"]["date_of_birth"] = datetime.fromisoformat(metadata["Subject"]["date_of_birth"])
-        nwbfile_kwargs.update(subject=Subject(**metadata["Subject"]))
-    # convert ISO 8601 string to datetime
-    if isinstance(nwbfile_kwargs.get("session_start_time", None), str):
-        nwbfile_kwargs["session_start_time"] = datetime.fromisoformat(metadata["NWBFile"]["session_start_time"])
-    return NWBFile(**nwbfile_kwargs)
+            source_data = session["source_data"]
+            for interface_name, interface_source_data in session["source_data"].items():
+                for key, value in interface_source_data.items():
+                    source_data[interface_name].update({key: str(Path(data_folder) / value)})
 
+            converter = CustomNWBConverter(source_data=source_data)
+            metadata = converter.get_metadata()
+            for metadata_source in [global_metadata, experiment_metadata, session.get("metadata", dict())]:
+                metadata = dict_deep_update(metadata, metadata_source)
 
-def check_regular_timestamps(ts):
-    """Check whether rate should be used instead of timestamps."""
-    time_tol_decimals = 9
-    uniq_diff_ts = np.unique(np.diff(ts).round(decimals=time_tol_decimals))
-    return len(uniq_diff_ts) == 1
+            print(metadata["NWBFile"])
+            if "nwbfile_name" not in session:
+                # 'subject_id' is required by schema validation if the 'Subject' is specified in metadata
+                assert "Subject" in metadata and "session_start_time" in metadata.get("NWBFile"), (
+                    "If not specifying an explicit name for the NWBFile ('nwbfile_name'), then both "
+                    "metadata['Subject']['subject_id'] and metadata['NWBFile']['session_start_time'] are required!"
+                )
+                subject_file_name = metadata["Subject"]["subject_id"].replace(" ", "_")
+                nwbfile_name = f"{subject_file_name}_{metadata['NWBFile']['session_start_time']}"
```

Comment:
could we use the dandi naming code here instead? https://github.com/dandi/dandi-cli/blob/master/dandi/organize.py#L70

Reply:
Yep, good idea